src/client/encode.rs

```diff
@@ -1,108 +1,106 @@
+            request,
+            state: EncoderState::Start,
         }
+    }
 
-        let val = format!("{} {} HTTP/1.1\r\n", req.method(), url);
-        log::trace!("> {}", &val);
-        buf.write_all(val.as_bytes()).await?;
-
-        if req.header(HOST).is_none() {
-            // Insert Host header
-            // Insert host
-            let host = req.url().host_str();
-            let host = host.ok_or_else(|| format_err!("Missing hostname"))?;
-            let val = if let Some(port) = req.url().port() {
-                format!("host: {}:{}\r\n", host, port)
+    fn finalize_headers(&mut self) -> io::Result<()> {
+        if self.request.header(HOST).is_none() {
+            let url = self.request.url();
+            let host = url
+                .host_str()
+                .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, "Missing hostname"))?
+                .to_owned();
+
+            if let Some(port) = url.port() {
+                self.request
+                    .insert_header(HOST, format!("{}:{}", host, port));
             } else {
-                format!("host: {}\r\n", host)
+                self.request.insert_header(HOST, host);
             };
-
-            log::trace!("> {}", &val);
-            buf.write_all(val.as_bytes()).await?;
         }
 
         // Insert Proxy-Connection header when method is CONNECT
-        if req.method() == Method::Connect {
-            let val = "proxy-connection: keep-alive\r\n".to_owned();
-            log::trace!("> {}", &val);
-            buf.write_all(val.as_bytes()).await?;
+        if self.request.method() == Method::Connect {
+            self.request.insert_header("proxy-connection", "keep-alive");
         }
 
         // If the body isn't streaming, we can set the content-length ahead of time. Else we need to
         // send all items in chunks.
-        if let Some(len) = req.len() {
-            let val = format!("content-length: {}\r\n", len);
-            log::trace!("> {}", &val);
-            buf.write_all(val.as_bytes()).await?;
+        if let Some(len) = self.request.len() {
+            self.request.insert_header(CONTENT_LENGTH, len.to_string());
         } else {
-            // write!(&mut buf, "Transfer-Encoding: chunked\r\n")?;
-            panic!("chunked encoding is not implemented yet");
-            // See: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding
-            //      https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Trailer
+            self.request.insert_header(TRANSFER_ENCODING, "chunked");
         }
 
-        for (header, values) in req.iter() {
-            for value in values.iter() {
-                let val = format!("{}: {}\r\n", header, value);
-                log::trace!("> {}", &val);
-                buf.write_all(val.as_bytes()).await?;
+        Ok(())
+    }
+
+    fn compute_head(&mut self) -> io::Result<Cursor<Vec<u8>>> {
+        let mut buf = Vec::with_capacity(100);
```

Comment:
Is this a guess size? 128 may be more efficient, since the next expansion for this Vec would probably go to 128 rather than 256.

Reply:
yeah, this was a ballpark guess. i'll up it to 128, seems like a nice number